### DISCUSSION TITLE
Moving away from hard coded, illogical numbers and decreasing choice

### DISCUSSION THREAD

**Discussion by royfalk** [vegastrike/Vega-Strike-Engine-Source#874] [General]
Issues
---------

**Abitrary numbers**
The game engine and especially VegaStrike is filled with arbitrary numbers.
This is mainly an issue with ships in VegaStrike. Llama has a mass of 250 and warp cost of 120. Plowshare has a mass of 800 and a warp cost of 120 too. Why not make warp cost a function of mass? Llama.begin the starting ship is actually a fairly heavy ship and can do with the lowest ftl capacitor. It shouldn't.
Also, the developers hard coded the reactor type (lithium6?) into the code and scientific units (MJ). 

**Wouldn't it be cool**
Wouldn't it be cool if we support cloak running on ftl capacitors instead of standard ones?
Wouldn't it be cool if a hit from behind went to the engines and one to the right side went to the radar?
The game code is filled with these. In the first example, the easiest and worst implementation means an if statement every loop.
In the second example, it turns a simple, random followed by a switch statement into something a lot more complicated.

**(Re)balancing**
Vegastrike works. But I bet someone who is more into the data than the engine knows several examples where the balance isn't quite right. With so many ships and components, all hard-coded, rebalancing is very very hard. Note that rebalance here doesn't mean ships need to be balanced. They just need to make sense somehow. 

**Llama.begin**
The starting ship deserves a section of its own. It's a good starting ship, but it somehow devolves into Llama.template/Llama.blank. With changes over time, this means that the originally correct (?) hard coded values are no longer. See #13 for one such example.

**Sensible Upgrades**
The same jump drive can power a fighter and a capital ship? Armor upgrade costs the same and weighs the same for fighters and capital ships?

Proposed Solutions
----------------------------

- Make the game rely on fewer parameters where possible. e.g. replace warp_cost with mass.
- Move the scientific units into python.
- Trim the wouldn't it be cool aggressively. Document it somewhere.
- Someone needs to take ownership of ships.json and rebalance. 
- Fix Llama.begin
- Implement the sensible upgrades as part of lib components.

**Comment by evertvorster** [comment]
Your solutions seem valid to me. 

To get more people playing the game, that Llama.begin is a pretty high priority. However, it is also one of the wouldn't it be cool examples. 
The idea that you can buy a ships hull and then upgrade it is really cool. 

But, if it can be done with one ship, why not make it possible on all ships? 
What makes the Llama that you get in the beginning so special that similar things can't be done on other ships?

The way the various special ships are implemented are really restrictive. You should be able to buy a special ship, and once it is in your inventory be able to modify any aspect of it just like you can with any other ship. 

So, the ship you buy is read off a template, but that template only sets the initial variables in what is forever more your ship. The game engine should read from the template only once, when you buy the ship. For the rest of the time, it should read from the save file. 
Then packages can be implemented better too. Say you buy a milspec package, it modifies certain parameters, but then the package itself is gone. You now just have the collection of parts that can be sold individually, and some modified parameters of your ship.

**Reply by royfalk** [reply]
> Llama.begin is a pretty high priority. 

Agree. 

> However, it is also one of the wouldn't it be cool examples. 

"wouldn't it be cool" is an idea that is marginal to the game play in general **and** complicates the code. 

>The idea that you can buy a ships hull and then upgrade it is really cool.

It is and it's staying. It's not what I meant by "wouldn't it be cool". Even if I wanted to, removing component support is a huge effort.

> What makes the Llama that you get in the beginning so special that similar things can't be done on other ships?

They can be done. Llama.begin is special because you get it at the start. If we want new players to get a flyable ship that's non-standard, it needs its own entry. Even if it only means the description at the top "with * death, it's now yours" or whatever is actually written there. 
 
> The way the various special ships are implemented are really restrictive. You should be able to buy a special ship, and once it is in your inventory be able to modify any aspect of it just like you can with any other ship.

It sorta works that way, but I've found that stuff is inconsistent. This is partly what lib component aims to fix.
 
> So, the ship you buy is read off a template, but that template only sets the initial variables in what is forever more your ship. 

That requires a change to the unit save file. You see, if something gets damaged or sold, the game checks against *.blank and *.template to figure out the correct values as these were never saved.

It's important to note that this discussion is really a way to get some consensus before implementing stuff in the code. I prefer not to remove certain features/code, only to discover that everyone else is against it. On the other hand, design documents or getting approval in advance is also a pain. @BenjamenMeyer is the only other developer and has a lot of other things to do.

**Reply by BenjamenMeyer** [reply]
I'd like to see us evolve the code to where things are easier to do and manage. It may take a bit of time to get there, and that's okay. Steps along the way with releasable/playable code will be needed to keep the community going.

And yeah I'm busy for the moment. But I do need to find some time to do stuff too. Right now have a certification test i need to do, after that time frees up a bit
